server/src/main/java/io/druid/discovery/DruidLeaderSelector.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.discovery;
+
+import javax.annotation.Nullable;
+
+/**
+ * Interface for supporting Overlord and Coordinator Leader Elections in TaskMaster and DruidCoordinator
+ * which expect appropriate implementation available in guice annotated with @IndexingService and @Coordinator
+ * respectively.
+ *
+ * Usage is as follow.
+ * On lifecycle start:
+ *  druidLeaderSelector.registerListener(myListener);
+ *  druidLeaderSelector.start();
+ *
+ * On lifecycle stop:
+ *  druidLeaderSelector.stop();
+ */
+public interface DruidLeaderSelector
+{
+
+  /**
+   * Get ID of current Leader.
+   */
+  @Nullable
+  String getCurrentLeader();
+
+  /**
+   * Returns true if this node is elected leader from underlying system's point of view. For example if curator
+   * is used to implement this then true would be returned when curator believes this node to be the leader.
+   */
+  boolean isLeader();
+
+  /**
+   * Implementation would increment it everytime it becomes leader. This allows users to start a long running
+   * task when they become leader and be able to intermittently check that they are still leader from same
+   * term when they started. DruidCoordinator class uses it to do intermittent checks and stop the activity
+   * as needed.
+   */
+  int localTerm();
+
+  /**
+   * Register the listener for watching leadership notifications.
+   */
+  void registerListener(Listener listener);
+
+  /**
+   * Must be called right after registerLeader(Listener).
+   */
+  void start();
```

Comment:
Maybe better to combine two above methods into a single method like ```registerListenerAndStart()```.

Reply:
ok, on second thought I removed start/stop methods altogether and instead have registerListener(listener) and unregisterListener() .

Reply:
Looks nice!